File: python/lesson 2/middle_node.py
from typing import Optional
from list_node import ListNode

# https://leetcode.com/problems/middle-of-the-linked-list/
class Solution:
    def middle_node(self, head: Optional[ListNode]) -> Optional[ListNode]:
        slow = fast = head
        while fast and fast.next:
            slow = slow.next
            fast = fast.next.next
        return slow

File: csharp/lesson 2/MiddleNode.cs
// https://leetcode.com/problems/middle-of-the-linked-list/

public class Solution 
{
    public ListNode MiddleNode(ListNode head)
    {
        ListNode slow = head, fast = head;
        while (fast?.next != null)
        {
            slow = slow.next;
            fast = fast.next.next;
        }
        return slow;
    }
}
